use allgatherv for sparse all reduce

zhaojuanmao opened the pull request:
Summary:
per #22226, The current sparse allreduce in ProcessGroupGloo pads the indices and values tensors to the maximum length across all processes and then performs a regular allgather (because they'll have equal size across processes). Instead, we can use allgatherv. This is mostly a win for memory usage if there is severe size imbalance between processes.

close #22226

Differential Revision: D16664985

pietern commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -755,29 +762,33 @@ class AsyncSparseAllreduceWork : public ProcessGroupGloo::AsyncWork {
-    buffer.select(0, context->rank)
-        .narrow(1, 0, tensor._nnz())
-        .copy_(tensor.indices());
+    auto output = at::empty({(int64_t)totalSize}, at::kLong);
```
Is the cast needed?

pietern commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -786,34 +797,47 @@ class AsyncSparseAllreduceWork : public ProcessGroupGloo::AsyncWork {
+      totalSize += counts[i];
+    }
+
+
```
Whitespace -- you can run `clang-format` to fix up style before committing.

pietern commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -755,29 +762,33 @@ class AsyncSparseAllreduceWork : public ProcessGroupGloo::AsyncWork {
     for (size_t i = 0; i < metadata.size(); i++) {
-      indices.push_back(buffer.select(0, i).narrow(1, 0, metadata[i].nnz()));
+      auto nnz = metadata[i].nnz();
+      auto sparseDim = tensor.sparse_dim();
```
This can be set as `const` outside the loop (and even at the beginning of the function).

pietern commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -786,34 +797,47 @@ class AsyncSparseAllreduceWork : public ProcessGroupGloo::AsyncWork {
-      max_nnz = std::max(max_nnz, metadata[i].nnz());
+    // There are nnz #dense_dim()-dimensional tensors per rank.
+    const auto valueShape = tensor.sizes().slice(tensor.sparse_dim());
+    size_t denseDims = 1;
```
This is not the number of dimensions but the number of elements in the dense dimensions, so wouldn't `denseNumel` (or similar) be more descriptive here?

pietern reviewed (commented):
Nice work! Mostly style/naming nits.

Also, I think that `setAllGatherVOutput` can be named just `setOutput` since the counts setter will be applicable to e.g. `gatherv`, `scatterv` as well.

pietern commented:
The test show something is wrong in the implementation.

pietern commented:
@zhaojuanmao It needs a rebase now because there seems to be a conflict with master.

Do you have time to continue this?

pietern commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -877,29 +883,38 @@ class AsyncSparseAllreduceWork : public ProcessGroupGloo::AsyncWork {
-    opts.setOutput(buffer.data_ptr<int64_t>(), buffer.numel());
+    // tensors copied from cuda may not be contiguous, get a contiguous
+    // tensor before use its data_ptr
+    auto input = tensor.indices().contiguous();
```
The inputs are always coalesced before running the algorithm and I was thinking that that implies they'll be contiguous.

Is this not the case?

pietern commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -877,29 +883,38 @@ class AsyncSparseAllreduceWork : public ProcessGroupGloo::AsyncWork {
     for (size_t i = 0; i < metadata.size(); i++) {
-      indices.push_back(buffer.select(0, i).narrow(1, 0, metadata[i].nnz()));
+      auto nnz = metadata[i].nnz();
+      auto numel = sparseDim * nnz;
```
These can both be const.

pietern commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -908,34 +923,47 @@ class AsyncSparseAllreduceWork : public ProcessGroupGloo::AsyncWork {
     for (size_t i = 0; i < metadata.size(); i++) {
-      values.push_back(buffer.select(0, i).narrow(0, 0, metadata[i].nnz()));
+      auto nnz = metadata[i].nnz();
+      auto numel = denseNumel * nnz;
```
These can both be const.

pietern commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -908,34 +923,47 @@ class AsyncSparseAllreduceWork : public ProcessGroupGloo::AsyncWork {
+    gloo::AllgathervOptions opts(context);
+    // tensors copied from cuda may not be contiguous, get a contiguous
+    // tensor before use its data_ptr
+    at::Tensor valueTensor = tensor.values().contiguous();
```
Same as above -- I don't think this is needed.

pietern reviewed (commented):
Can you clarify the need for the `contiguous()` call?

Don't worry about the const-ness in the loop body if this doesn't require further changing.

zhaojuanmao commented:
@pietern some tensors copied from cuda are not be contiguous, this is what I printed out locally while debugging, If I directly print out values using tensor data pointer, some values are garbage values.

pietern commented:
This was against tensors when running tests? Do you have a repro for this?

I'm surprised this was not an issue before, because the `max_nnz` padding may have had the same issue and caused garbage crashes as well.

zhaojuanmao commented:
this is same for existing codes, I printed out locally for existing codes, some tensors copies from cuda is not contiguous. but they called copy_ to a contiguous buffer tensors before calling allgather, the copy_ is similar to call contiguous() call

zhaojuanmao commented:
yes, it is for tensors in unit tests. I printed values and is_contiguous() flag, for tensors with garbage values using data pointer, is_contiguous() is false

pietern commented:
Ah yes, of course!

Can you create an issue for this? We call coalesce, which gives the underlying code the opportunity to coalesce as well. I don't think there is a good reason for not creating indices/values in an non-coalesced tensor if you're creating it anyway, so perhaps this is a legit bug.

Let's merge this now and then update if this is indeed a bug and there is a resolution.

zhaojuanmao commented:
> Ah yes, of course!
> 
> Can you create an issue for this? We call coalesce, which gives the underlying code the opportunity to coalesce as well. I don't think there is a good reason for not creating indices/values in an non-coalesced tensor if you're creating it anyway, so perhaps this is a legit bug.
> 
> Let's merge this now and then update if this is indeed a bug and there is a resolution.

asked whether it is expected to have non-contiguous indices tensor after calling coalesce() for the sparse tensor here https://discuss.pytorch.org/t/indices-tensor-is-not-contiguous-after-calling-sparse-tensor-coalesce/56198, because based on codes https://our.internmc.facebook.com/intern/diffusion/FBS/browse/master/fbcode/caffe2/aten/src/ATen/native/sparse/SparseTensor.cpp?lines=350, looks like indices tensor should be contiguous

if the answer is not expected, will create a bug PR issue

zhaojuanmao commented:
@pytorchbot retest

pietern commented:
The "ensure no tabs" linter is for unrelated changes on master (which should be fixed separately).

zhaojuanmao commented:
@pietern yesterday MacOS failed to build because it can not recognize "long" symbols somehow, I changed it to int64_t, now it passed the tests. 

will land it soon

zhaojuanmao committed:
use allgatherv for sparse all reduce (#23917)

Summary:
per https://github.com/pytorch/pytorch/issues/22226, The current sparse allreduce in ProcessGroupGloo pads the indices and values tensors to the maximum length across all processes and then performs a regular allgather (because they'll have equal size across processes). Instead, we can use allgatherv. This is mostly a win for memory usage if there is severe size imbalance between processes.

close https://github.com/pytorch/pytorch/issues/22226
Pull Request resolved: https://github.com/pytorch/pytorch/pull/23917

Test Plan:
buck run mode/dev-nosan caffe2/test:c10d -- test_c10d.ProcessGroupGlooTest.test_sparse_allreduce_basics

buck run mode/dev-nosan caffe2/test:c10d -- test_c10d.ProcessGroupGlooTest.test_sparse_allreduce_basics_cuda

buck run mode/dev-nosan caffe2/test:c10d -- test_c10d.ProcessGroupGlooTest.test_sparse_allreduce_checks

Differential Revision: D16664985

Pulled By: zhaojuanmao

fbshipit-source-id: a9a139da2b64617d2bb7f0b12f272e920140e5d1
